source/custom/nixnet_service.custom.cpp

```diff
@@ -1,2 +1,292 @@
+        }
+        case nixnet_grpc::ReadState::READ_STATE_STATE_CAN_COMM :
+        case nixnet_grpc::ReadState::READ_STATE_STATE_FLEX_RAY_COMM :
+        case nixnet_grpc::ReadState::READ_STATE_STATE_SESSION_INFO :{
+            state_size = sizeof(u32);
+            break;
+        }
+        case nixnet_grpc::ReadState::READ_STATE_STATE_FLEX_RAY_STATS :{
+            state_size = sizeof(nxFlexRayStats_t);
+            break;
+        }
+        case nixnet_grpc::ReadState::READ_STATE_STATE_LIN_COMM :{
+            // The StateValue for nxState_LinComm should point to a u32 array with 2 elements.
+            state_size = sizeof(u32)*2;
+            break;
+        }
+        case nixnet_grpc::ReadState::READ_STATE_STATE_J1939_COMM :{
+            state_size = sizeof(nxJ1939CommState_t);
+            break;
+        }
+        case nixnet_grpc::ReadState::READ_STATE_STATE_TIME_CURRENT_2 : 
+        case nixnet_grpc::ReadState::READ_STATE_STATE_TIME_COMMUNICATING_2 : 
+        case nixnet_grpc::ReadState::READ_STATE_STATE_TIME_START_2 :{
+            state_size = sizeof(nxTimeLocalNetwork_t);
+            break;
+        }
+        default:{
+            return ::grpc::INVALID_ARGUMENT;
+            break;
+        }
+    }
+        return state_size;
+}
+
+// ReadState API has an output parameter of type void * called StateValue. Based on the value of StateID, 
+// StateValue can point to u32, nxTimestamp100ns_t,  _nxFlexRayStats_t etc, which are of different sizes.
+// Based on the StateID, we are setting the size of StateValue and after calling the ReadState API, the 
+// response is set appropriately.
+::grpc::Status NiXnetService::ReadState(::grpc::ServerContext* context, const ReadStateRequest* request, ReadStateResponse* response)
+{
+    if (context->IsCancelled()) {
+        return ::grpc::Status::CANCELLED;
+    }
+    try {
+        auto session_ref_grpc_session = request->session_ref();
+        nxSessionRef_t session_ref = session_repository_->access_session(session_ref_grpc_session.id(), session_ref_grpc_session.name());
+        u32 state_id;
+        switch (request->state_id_enum_case()) {
+            case nixnet_grpc::ReadStateRequest::StateIdEnumCase::kStateId: {
+                state_id = static_cast<u32>(request->state_id());
+                break;
+            }
+            case nixnet_grpc::ReadStateRequest::StateIdEnumCase::kStateIdRaw: {
+                state_id = static_cast<u32>(request->state_id_raw());
+                break;
+            }
+            case nixnet_grpc::ReadStateRequest::StateIdEnumCase::STATE_ID_ENUM_NOT_SET: {
+                return ::grpc::Status(::grpc::INVALID_ARGUMENT, "The value for state_id was not specified or out of range");
+                break;
+            }
+        }
+
+        u32 state_size = GetStateSize(state_id);
+        // Since GetStateSize returns only u32, in case of error, we are returning a meaningful message
+        if (state_size == ::grpc::INVALID_ARGUMENT)
```

Comment:
I think you might need to throw something from GetStateSize and have a Catch to return this Status. As is, this prevents having a state_size of 3: https://grpc.github.io/grpc/core/md_doc_statuscodes.html